impress-org/givewp
# Allow custom amount field in Donor Dasboard Subscription Management to be emptied

Resolves #5896

## Description



In the subscription management view of the donor dashboard, there is an input that appears when the user chooses "Custom Amount." This input previously would replace its empty state with the minimum allowed value. This seems like it is desirable, however, it is not. In fact, users without the knack of filling out forms on the Internet will find it impossible to change the value to an amount where the most significant digit is less than the most significant digit in the minimum value (e.g. the minimum is $5 and the user wants to update the value to $12, but needs to change it from $15, once they backspace once, so that the input $reads 1, the input will think the value is less than $5 and change it to $5, since $1 is less than $5, even though they intend to write $12).

This PR resolves this issue by not setting minimum and maximum values when the value in the input is outside of them. Instead, we simply validate the input on blur and provide the user the appropriate feedback they need to enter an acceptable amount.

The existing logic in these components was so hard to follow and, since I had redefined the functionality, I completely scrapped what was there. The implementation that I needed was far simpler and I reduced the complexity further by normalizing the amount, the minimum, and the maximum _before_ giving them to the controls. This means that the underlying controls work in a more controlled manner.

This also gets rid of the `CurrencyControl` component. This was the only view that was using it, so I think it was a premature abstraction. We can better think through a re-usable version in the future.

## Affects



- Subscription Management view of the Donor Dashboard.

## Visuals



https://user-images.githubusercontent.com/6908001/135361310-2ba270b2-3827-4d66-b7f8-29b0477fe0e4.mp4

## Testing Instructions



Requirements:

- Recurring addon is installed and activated.
- Make sure you are using the Stripe gateway.

Instructions:

1. Set up a form that allows recurring donations.
2. Make a recurring donation.
3. View the donation dashboard for the donor (`/donor-dashboard`).
4. Go to "Recurring Donations."
4. Press "Manage Subscription" on the recurring donation.
5. Set the select to "Custom Amount."
6. Play around with setting the value correctly and incorrectly.

In general you might want to test the whole subscription management feature out since I changed a fair bit of code there.

## Pre-review Checklist
